### PR TITLE
Added preserve_tree option to the config file.

### PR DESCRIPTION
**Pull request by chintal**
This adds the following options to the config file, along with the ability to optionally change the behaviour of the default output directory. 

- preserve_tree
- server_roots

Usage example : 

    [some-name]
    vcs = svn
    basedir = /some/directory
    preserve_tree = True
    server_roots = 
            svn://some.svn.server/
    checkouts =
            svn://some.svn.server/folder1/folder1a/repo1
            svn://some.svn.server/folder1/folder1b/repo1
            svn://some.svn.server/folder2/folder2a/repo2
            svn://some.svn.server/folder3/folder3c/repo5

Which creates the following tree : 

    /some/directory/folder1/folder1a/repo1
    /some/directory/folder1/folder1b/repo1
    /some/directory/folder2/folder2a/repo2
    /some/directory/folder3/folder3c/repo5

### CODE REVIEW

**Comment by reinout** [comment]
Very quick because of a meeting: can you merge the two options into one? Something like:

    preserve_root = svn://some.server

Less options, especially since they belong together.

**Comment by chintal** [comment]
Done. Following is the updated example. 

    [some-name]
    vcs = svn
    basedir = /some/directory
    preserve_tree = 
            svn://some.svn.server/
    checkouts =
            svn://some.svn.server/folder1/folder1a/repo1
            svn://some.svn.server/folder1/folder1b/repo1
            svn://some.svn.server/folder2/folder2a/repo2
            svn://some.svn.server/folder3/folder3c/repo5

File: checkoutmanager/config.py [hunk 1]
```diff
@@ -33,22 +36,27 @@ def extract_spec(spec):
...
-        parts = [part for part in vcs_url.split('/')
-                 if part]
-        # Common structure: having a customer folder with a 'buildout'
-        # directory in it.  Don't name it 'buildout'.
-        parts = [part for part in parts if part != 'buildout']
-        directory = parts[-1]
-        # Remove /trunk from the end.  We don't want that as a name.
-        if parts[-1] == 'trunk':
-            parts.pop()
+        for sroot in server_roots:
```
Comment (reinout) [inline checkoutmanager/config.py]: Don't abbreviate, just call it `server_root` instead of `sroot` :-)

File: README.rst [hunk 2]
```diff
@@ -116,6 +116,23 @@ takes the last part.  Some exceptions, most for subversion:
 
 - Git checkouts lose the ".git" at the end of the url.
 
+Optionally, a "preserve_tree" option may be set for each group in the
+configuration. The contents of the "preserve_tree" option should be a list
+of strings. If the "preserve_tree" option is set, then the default
```
Comment (reinout) [inline README.rst]: I wouldn't emphasise the "list"-ness too much. And "list" and "string" are so datatype-like. What about: "In `preserve_tree` you can set one or more base checkout urls."?

Ah well, leave that in for now. I'll make it a bit shorter myself, I've got some ideas :-) Before making a release I'll let you see it if you think it is still clear enough.

**Comment by reinout** [comment]
I added one inline comment.

I've got some small things I want to change in the documentation, but I'll do that myself and show you afterards.

Could you also add a quick test in the "Determine vcs url and directory name" section in, I think, `tests/config.txt`?

**Comment by reinout** [comment]
Oh, and a short note in the `CHANGES.rst` :-)

**Comment by chintal** [comment]
I will try to do so. I'll have to think up the test cases somewhat - it's been a while since I've actually written any unit tests.

**Comment by reinout** [comment]
Or I can add one, that helps me check if the code is correct now.

**Comment by reinout** [comment]
If you want me to add the test, can you update the changelog and warn me when ready? Then I can merge it.

**Comment by chintal** [comment]
I'll give it a go and let you know.

**Comment by reinout** [comment]
No hurry. I'm off to bed in a few minutes so it'll have to wait till tomorrow in any case :-)

**Comment by reinout** [comment]
Can you look at #12? I've added two extra tests that test corner cases. The code didn't need changing, but the tests provide extra security.

I also included preserve_tree in the regular set of naming rules in the readme.
